src/System.Drawing.Common/src/System/Drawing/Printing/PrintingServices.Unix.cs

```diff
@@ -589,7 +605,12 @@ private static PrinterResolution ParseResolution(string resolution)
         /// </summary>
         private static void LoadPrinters()
         {
-            installed_printers.Clear();
+            lock (installed_printers)
+            {
+                installed_printers.Clear();
+                populatedPrinters = false;
```

Comment:
@stephentoub Stephen, what do you think about the approach of making a task out of loadPrinter and storing this task with the static class. In this case, I will check if the task is currently running on each call and wait for it to end. Otherwise, I can start the task.

Reply:
@cojacfar I'm just arriving at this issue but this sounds like exactly what `Lazy<T>`  is designed for: https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization?view=netframework-4.8

Reply:
> Stephen, what do you think about the approach

Seems like LazyInitializer.EnsureInitialized would suit this well.  Does it not?  Clearing would just be matter of setting the field to null, and EnsureInitialized would take a ref to the field and the delegate that constructs the list of printers and returns it.

Reply:
So I turned it into this:

LazyInitializer.EnsureInitialized(ref installed_printers, () => { installed_printers = new Hashtable(); // Loads values into installed_printers LoadPrinters(); // Creates list from installed_printers that we return PrinterSettings.StringCollection list = new PrinterSettings.StringCollection(Array.Empty<string>()); foreach (object key in installed_printers.Keys) { list.Add(key.ToString()); } // Setting installed_printers back to null to cause EnsureInitialized to be called again after these threads finish // Also need this to dispose of the concurrently created hashtables installed_printers = null; return list; });

If I understand correctly, this should work. I had to drop the installed_printers creation in the constructor, which seems to have triggered a merge conflict.

Each thread is going to make a new Hashtable, which is a downside. I could instead create a new property to just cause re-initialization. Thinking about it, that might be the best way.

Reply:
I'm sorry if I mislead you somehow, but I don't understand what this is doing.  My suggestion was that the loading of the printers and the publishing of the relevant loaded data be done with LazyInitializer.EnsureInitialized, e.g.
```C#
private static Hashtable s_installedPrinters;

// When you need to read from the installed printers
Hashtable printers = LazyInitializer.EnsureInitialized(ref s_installedPrinters, () => 
{
    var printers = new Hashtable();
   ... // load all printers and store into printers
    return printers;
});
...
// When you need to clear the installed printers
Volatile.Write(ref s_installedPrinters, null);
...